File: dotnet/cs/src/Transducer.cs
/*
 * File: Transducer.cs
 * Project: Internal
 * Created Date: 08/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 06/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */


#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
#endif

using AUTD3Sharp.NativeMethods;

#if UNITY_2018_3_OR_NEWER
using UnityEngine;
using Vector3 = UnityEngine.Vector3;
using Quaternion = UnityEngine.Quaternion;
#else
using Vector3 = AUTD3Sharp.Utils.Vector3d;
using Quaternion = AUTD3Sharp.Utils.Quaterniond;
#endif

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp
{
    public sealed class Transducer
    {
        internal readonly TransducerPtr Ptr;

        internal Transducer(int trIdx, DevicePtr ptr)
        {
            Idx = trIdx;
            Ptr = NativeMethodsBase.AUTDTransducer(ptr, (uint)trIdx);
        }

        /// <summary>
        /// Index of the transducer
        /// </summary>
        public int Idx { get; }

        /// <summary>
        /// Position of the transducer
        /// </summary>
        public Vector3 Position
        {
            get
            {
                unsafe
                {
                    float_t* pos = stackalloc float_t[3];
                    NativeMethodsBase.AUTDTransducerPosition(Ptr, pos);
                    return new Vector3(pos[0], pos[1], pos[2]);
                }
            }
        }

        /// <summary>
        /// Rotation of the transducer
        /// </summary>
        public Quaternion Rotation
        {
            get
            {
                unsafe
                {
                    float_t* rot = stackalloc float_t[3];
                    NativeMethodsBase.AUTDTransducerRotation(Ptr, rot);
                    return new Quaternion(rot[1], rot[2], rot[3], rot[0]);
                }
            }
        }

        /// <summary>
        /// X-direction of the transducer
        /// </summary>
        public Vector3 XDirection
        {
            get
            {
                unsafe
                {
                    float_t* dir = stackalloc float_t[3];
                    NativeMethodsBase.AUTDTransducerDirectionX(Ptr, dir);
                    return new Vector3(dir[0], dir[1], dir[2]);
                }
            }
        }

        /// <summary>
        /// Y-direction of the transducer
        /// </summary>
        public Vector3 YDirection
        {
            get
            {
                unsafe
                {
                    float_t* dir = stackalloc float_t[3];
                    NativeMethodsBase.AUTDTransducerDirectionY(Ptr, dir);
                    return new Vector3(dir[0], dir[1], dir[2]);
                }
            }
        }

        /// <summary>
        /// Z-direction of the transducer
        /// </summary>
        public Vector3 ZDirection
        {
            get
            {
                unsafe
                {
                    float_t* dir = stackalloc float_t[3];
                    NativeMethodsBase.AUTDTransducerDirectionZ(Ptr, dir);
                    return new Vector3(dir[0], dir[1], dir[2]);
                }
            }
        }

        /// <summary>
        /// Wavelength of the transducer
        /// </summary>
        /// <param name="soundSpeed">Speed of sound</param>
        /// <returns></returns>
        public float_t Wavelength(float_t soundSpeed) => NativeMethodsBase.AUTDTransducerWavelength(Ptr, soundSpeed);

        /// <summary>
        /// Wavenumber of the transducer
        /// </summary>
        /// <param name="soundSpeed">Speed of sound</param>
        /// <returns></returns>
        public float_t Wavenumber(float_t soundSpeed) => 2 * AUTD3.Pi / Wavelength(soundSpeed);
    }
}

File: cpp/tests/internal/geometry/transducer.cpp
// File: transducer.cpp
// Project: geometry
// Created Date: 26/09/2023
// Author: Shun Suzuki
// -----
// Last Modified: 02/12/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include <gtest/gtest.h>

#include <autd3/internal/geometry/transducer.hpp>
#include <ranges>

#include "utils.hpp"

TEST(Internal_Geometry, TransducerIdx) {
  for (auto autd = create_controller(); auto& dev : autd.geometry()) {
    std::ranges::for_each(std::views::iota(0) | std::views::take(dev.num_transducers()), [&dev](auto i) { ASSERT_EQ(dev[i].idx(), i); });
  }
}

TEST(Internal_Geometry, TransducerPosition) {
  auto autd = create_controller();

  ASSERT_EQ(autd.geometry()[0][0].position(), autd3::internal::Vector3(0, 0, 0));
  ASSERT_EQ(autd.geometry()[0][autd3::internal::geometry::AUTD3::NUM_TRANS_IN_UNIT - 1].position(), autd3::internal::Vector3(172.72, 132.08, 0));

  ASSERT_EQ(autd.geometry()[1][0].position(), autd3::internal::Vector3(0, 0, 0));
  ASSERT_EQ(autd.geometry()[1][autd3::internal::geometry::AUTD3::NUM_TRANS_IN_UNIT - 1].position(), autd3::internal::Vector3(172.72, 132.08, 0));
}

TEST(Internal_Geometry, TransducerRotation) {
  for (auto autd = create_controller(); auto& dev : autd.geometry()) {
    std::ranges::for_each(dev.transducers(), [](auto& tr) { ASSERT_EQ(tr.rotation(), autd3::internal::Quaternion::Identity()); });
  }
}

TEST(Internal_Geometry, TransducerDirectionX) {
  for (auto autd = create_controller(); auto& dev : autd.geometry()) {
    std::ranges::for_each(dev.transducers(), [](auto& tr) { ASSERT_EQ(tr.x_direction(), autd3::internal::Vector3::UnitX()); });
  }
}

TEST(Internal_Geometry, TransducerDirectionY) {
  for (auto autd = create_controller(); auto& dev : autd.geometry()) {
    std::ranges::for_each(dev.transducers(), [](auto& tr) { ASSERT_EQ(tr.y_direction(), autd3::internal::Vector3::UnitY()); });
  }
}

TEST(Internal_Geometry, TransducerDirectionZ) {
  for (auto autd = create_controller(); auto& dev : autd.geometry()) {
    std::ranges::for_each(dev.transducers(), [](auto& tr) { ASSERT_EQ(tr.z_direction(), autd3::internal::Vector3::UnitZ()); });
  }
}

TEST(Internal_Geometry, TransducerWavelength) {
  for (auto autd = create_controller(); auto& dev : autd.geometry()) {
    for (auto& tr : dev) {
      ASSERT_DOUBLE_EQ(tr.wavelength(340e3), 340e3 / autd3::internal::native_methods::ULTRASOUND_FREQUENCY);
    }
  }
}

TEST(Internal_Geometry, TransducerWavenum) {
  for (auto autd = create_controller(); auto& dev : autd.geometry()) {
    for (auto& tr : dev) {
      ASSERT_DOUBLE_EQ(tr.wavenumber(340e3), 2 * autd3::internal::pi * autd3::internal::native_methods::ULTRASOUND_FREQUENCY / 340e3);
    }
  }
}
